module/component/component.go

```diff
@@ -102,11 +103,9 @@ func RunComponent(ctx context.Context, componentFactory ComponentFactory, handle
 			return err // failure to start
 		}
 
-		select {
-		case <-ctx.Done():
-			stop()
-			return ctx.Err()
-		case err := <-irrecoverableErr:
+		doneCtx, _ := util.WithDone(ctx, done)
```

Comment:
So much cleaner

Reply:
I think for `WaitError` though, since we don't do anything with the context error (as opposed to `WaitClosed` where we return the context error), we can just take in a done channel instead of the entire context.